File: k2/python/host/k2host/aux_labels.py
# Copyright      2020  Xiaomi Corporation (author: Haowen Qiu)

# See ../../../LICENSE for clarification regarding multiple authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
from torch.utils.dlpack import to_dlpack

from _k2host import IntArray2Size
from _k2host import _AuxLabels1Mapper
from _k2host import _AuxLabels2Mapper
from _k2host import _FstInverter

from .fsa import Fsa
from .array import IntArray1
from .array import IntArray2

AuxLabels = IntArray2


class AuxLabels1Mapper(_AuxLabels1Mapper):

    def __init__(self, labels_in: AuxLabels, arc_map: IntArray1):
        super().__init__(labels_in.get_base(), arc_map.get_base())

    def get_sizes(self, aux_size: IntArray2Size) -> None:
        return super().get_sizes(aux_size)

    def get_output(self, labels_out: AuxLabels) -> None:
        return super().get_output(labels_out.get_base())


class AuxLabels2Mapper(_AuxLabels2Mapper):

    def __init__(self, labels_in: AuxLabels, arc_map: IntArray2):
        super().__init__(labels_in.get_base(), arc_map.get_base())

    def get_sizes(self, aux_size: IntArray2Size) -> None:
        return super().get_sizes(aux_size)

    def get_output(self, labels_out: AuxLabels) -> None:
        return super().get_output(labels_out.get_base())


class FstInverter(_FstInverter):

    def __init__(self, fsa_in: Fsa, labels_in: AuxLabels):
        super().__init__(fsa_in.get_base(), labels_in.get_base())

    def get_sizes(self, fsa_size: IntArray2Size,
                  aux_size: IntArray2Size) -> None:
        return super().get_sizes(fsa_size, aux_size)

    def get_output(self, fsa_out: Fsa, labels_out: AuxLabels) -> None:
        return super().get_output(fsa_out.get_base(), labels_out.get_base())

File: k2/python/host/csrc/aux_labels.cc
// k2/python/host/csrc/aux_labels.cc

// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

// See ../../../LICENSE for clarification regarding multiple authors

#include "k2/python/host/csrc/aux_labels.h"

#include "k2/csrc/host/aux_labels.h"

void PyBindAuxLabels1Mapper(py::module &m) {
  using PyClass = k2host::AuxLabels1Mapper;
  py::class_<PyClass>(m, "_AuxLabels1Mapper")
      .def(py::init<const k2host::AuxLabels &,
                    const k2host::Array1<int32_t *> &>(),
           py::arg("labels_in"), py::arg("arc_map"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("aux_size"))
      .def("get_output", &PyClass::GetOutput, py::arg("labels_out"));
}

void PyBindAuxLabels2Mapper(py::module &m) {
  using PyClass = k2host::AuxLabels2Mapper;
  py::class_<PyClass>(m, "_AuxLabels2Mapper")
      .def(py::init<const k2host::AuxLabels &,
                    const k2host::Array2<int32_t *> &>(),
           py::arg("labels_in"), py::arg("arc_map"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("aux_size"))
      .def("get_output", &PyClass::GetOutput, py::arg("labels_out"));
}

void PyBindFstInverter(py::module &m) {
  using PyClass = k2host::FstInverter;
  py::class_<PyClass>(m, "_FstInverter")
      .def(py::init<const k2host::Fsa &, const k2host::AuxLabels &>(),
           py::arg("fsa_in"), py::arg("labels_in"))
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"),
           py::arg("aux_size"))
      .def("get_output", &PyClass::GetOutput, py::arg("fsa_out"),
           py::arg("labels_out"));
}

void PybindAuxLabels(py::module &m) {
  PyBindAuxLabels1Mapper(m);
  PyBindAuxLabels2Mapper(m);
  PyBindFstInverter(m);
}
